homeassistant/components/shelly/__init__.py

```diff
@@ -117,17 +124,18 @@ async def async_setup(self):
             sw_version=self.device.settings["fw"],
         )
 
-    async def shutdown(self):
+    async def shutdown(self, ha_stop=False):
         """Shutdown the device wrapper."""
         if self._unsub_stop:
             self._unsub_stop()
             self._unsub_stop = None
-        await self.device.shutdown()
+        if ha_stop and COAP_CONTEXT in self.hass.data[DOMAIN]:
```

Comment:
Instead of this, we should subscribe to the stop event when we create the context. That way we keep all the lifecycle of the COAP context in 1 place instead of relying on our device wrapper to clean up.